Move `contrib` to `root/opt/app-root` and update common

torsava opened the pull request:
common-container-scripts new automatically compile the README.md file into a
help.1 and place it into the root subdirectory to be copied to the image during
build.

I've separated the update to the Fedora Dockerfile to it's own commit, since it was suspiciously missing the instruction to copy the copy directory into the image in the first place. Two of the files therein were copied one by one, but the remaining file was not. I added the COPY instruction to unify it with the other Dockerfiles.

torsava committed:
Move `contrib` to `root/opt/app-root` and update common

common-container-scripts new automatically compile the README.md file into a
help.1 and place it into the root subdirectory to be copied to the image during
build.

torsava committed:
Move `contrib` to `root/opt/app-root` in Fedora Dockerfile

common-container-scripts new automatically compile the README.md file into a
help.1 and place it into the root subdirectory to be copied to the image during
build.

omron93 commented on 4-perftools/Dockerfile:
```diff
@@ -33,7 +32,7 @@ USER 0
 WORKDIR ${HOME}
 
-COPY contrib/etc/scl_enable /root/etc/scl_enable
+COPY root/root/etc/scl_enable /root/etc/scl_enable
```
@torsava Why `root/root`?

torsava replied:
The first `root` is the directory we have in this git repo that contains all the files that should be copied to `/`. And the second `root` is because the contents of the previous `contrib` directory were actually copied into `/root` (see above).

omron93 replied:
You are right. Thank you for explanation.

omron93 commented on 4-perftools/Dockerfile.rhel7:
```diff
@@ -38,7 +37,7 @@ USER 0
 WORKDIR ${HOME}
 
-COPY contrib/etc/scl_enable /root/etc/scl_enable
+COPY root/root/etc/scl_enable /root/etc/scl_enable
```
This is not needed IMPOV - `scl_enable` is already in `/root/etc/scl_enable` by `COPY ./root/ /`

torsava replied:
I also believe it's not necessary (though it's after the `USER 0` declaration which might affect it?), but I think that it should be removed as part of a different PR so as not to muddy the history.

omron93 replied:
`USER 0` is default, so this command is redundant too and might not affect the COPY.

>  I think that it should be removed as part of a different PR so as not to muddy the history.

OK. I thought it could be, because it related to getting files into container. And aim of `root/` folder is to be able to add all files in once (in almost all cases).

torsava replied:
> USER 0 is default, so this command is redundant too and might not affect the COPY.

I was thinking maybe it has an effect on some obscure container platform, otherwise the instruction doesn't make sense. But I agree it's most likely just an oversight. I'll add a separate commit to get rid of the superfluous copies.

omron93 commented on 4-toolchain/Dockerfile:
```diff
@@ -47,12 +46,12 @@ WORKDIR ${HOME}
 # Use entrypoint so path is correctly adjusted already at the time the command
 # is searching, so something like docker run IMG gcc runs binary from SCL.
-COPY contrib/bin/container-entrypoint /usr/bin/container-entrypoint
+COPY root/opt/app-root/bin/container-entrypoint /usr/bin/container-entrypoint
```
The same - why not to add `container-entrypoint` into `root/usr/bin/` and don't copy it separately?

@hhorak Do you know why it is needed in `/usr/bin/` (because `/opt/app-root/bin/` is also in PATH) ? Is it necessary to have `container-entrypoint` in both locations now? (same about `usage` script)

torsava replied:
If there is a reason to have it in both location, we could just make it a symlink inside the `root` subdirectory in this repo so that it gets copied to the image without needing a separate Docker instruction.

hhorak replied:
It seems `/usr/bin/` is used in other containers for `container-entrypoint`.

torsava commented on 4-toolchain/Dockerfile.rhel7:
```diff
@@ -52,12 +51,12 @@ WORKDIR ${HOME}
+COPY root/opt/app-root/bin/usage /usr/local/bin/usage
 
-COPY contrib/etc/scl_enable /opt/app-root/etc/scl_enable
+COPY root/opt/app-root/etc/scl_enable /opt/app-root/etc/scl_enable
```
@omron93 What about here? It's re-copied after the `USER 1001` instruction. I don't think we need `scl_enable` owned by the user.

torsava replied:
In addition, it's after the instruction

    chown -R 1001:1001 /opt/app-root

Therefore it already *is* owned by the user. Removing.

omron93 replied:
Maybe we needs. Or correct permissions have to be set.

So change `chmod u+x /opt/app-root/bin/usage` to `chmod a+rx /usr/bin/usage /usr/bin/container-entrypoint` (or `/opt/app-root/bin` if these files should be there) ?

torsava replied:
Here I was talking only about the highlighted `/opt/app-root-etc/scl_enable` COPY instruction, which can be safely removed.

With `usage` and `container-entrypoint`, we have to decide if we need them at `/usr/(local/)bin` and if so, whether to do it using a Docker instruction like it's already done, or maybe a symlink inside the `root` subdirectory of this repo.

torsava committed:
Remove superfluous COPY instructions

torsava commented:
I added a commit that removes the superfluous COPY instructions. They had no effect since the files already were in the desired location and were owned by the desired user.

torsava committed:
Deduplicate the locations of container-entrypoint and usage

They were previously copied to `/opt/app-root-bin` and then copied again into
`/usr/(local/)bin/`. They have now been moved to the second location only.

torsava commented:
I've added a commit that deduplicates the location of `container-entryopoint` and `usage` scripts.

torsava commented:
Fixed a typo that caused the failed tests.

hhorak commented:
Looks good to me. I'd also change location of scripts (`usage`) from `/usr/local/bin` to `/usr/bin`. I'm fine with doing it in this PR or opening a separate PR.. no preference on my side..

torsava committed:
Move usage script from /usr/local/bin to /usr/bin

torsava commented:
@hhorak I've moved it in the latest commit.

omron93 commented:
@torsava Please move `usage` and `container-entrypoint` also in `{4,6}-perftools` images.

torsava committed:
Moving `usage` and `container-entrypoint` from `/root/bin/` to `/usr/bin` in *-perftools

torsava commented:
@omron93 Moved both scripts from `/root/bin/` to `/usr/bin/` in `*-perftools` to be in line with `*-toolchain`.

omron93 commented:
Thanks.

LGTM + tests are passing. So merging.

torsava commented:
@omron93 Thank you for all the reviews and merges on this voyage :)
